File: skills/agent-cli-builder/templates/python-typer/src/mycli/async_tasks.py
"""Async task pattern — invariant #11.

Anything that takes longer than ~5 seconds should not block the agent.
The skeleton here is intentionally minimal: a `Task` dataclass, a
`TaskStore` Protocol, and a `wait_for` helper.

What's NOT in this template by default:
* a concrete file/SQLite/HTTP `TaskStore` implementation
* `cancel` / `list` / `download`
* task creation flows

Those are domain-specific and easy to write once you know the contract.
See `templates/RECIPES.md` in the parent agent-cli-builder skill for
worked examples (file-backed store with cancel + list, download).
"""
from __future__ import annotations

import time
from dataclasses import asdict, dataclass, field
from typing import Any, Literal, Protocol

from .errors import TimeoutCliError

TaskState = Literal["pending", "running", "succeeded", "failed", "cancelled"]
_TERMINAL: tuple[TaskState, ...] = ("succeeded", "failed", "cancelled")


@dataclass
class Task:
    task_id: str
    kind: str
    state: TaskState = "pending"
    progress: float = 0.0
    created_at: float = field(default_factory=time.time)
    updated_at: float = field(default_factory=time.time)
    eta_seconds: float | None = None
    result: Any = None
    error: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


class TaskStore(Protocol):
    """The task storage contract. Implement against your real backend.

    `mycli task get` and `mycli task wait` only need `get`. Add `cancel`
    and `list` when you need them; see the recipes file for examples.
    """

    def get(self, task_id: str) -> Task: ...


def wait_for(
    store: TaskStore,
    task_id: str,
    *,
    timeout_seconds: float = 300.0,
    poll_interval_seconds: float = 1.0,
) -> Task:
    """Block until the task reaches a terminal state, or timeout.

    Returns the final task on success; raises `TimeoutCliError` (exit 5)
    on timeout so agents can branch on `$?` without parsing strings.
    """
    deadline = time.time() + timeout_seconds
    while True:
        task = store.get(task_id)
        if task.state in _TERMINAL:
            return task
        if time.time() > deadline:
            raise TimeoutCliError(
                f"timed out after {timeout_seconds}s waiting for {task_id}",
                suggestions=[
                    f"Poll with `mycli task get {task_id}` to check state.",
                    "Increase --timeout if the task legitimately takes longer.",
                ],
            )
        time.sleep(poll_interval_seconds)

File: skills/agent-cli-builder/templates/rust-clap/crates/mycli-core/src/async_tasks.rs
//! The async task pattern — invariant #11.
//!
//! Anything that takes longer than ~5 seconds should not block the agent.
//! The skeleton is intentionally minimal: a `Task` value type, a
//! `TaskStore` trait, and a `wait_for_terminal` helper. Both `get` and
//! `wait` are wired in `mycli-cli/src/commands/task.rs`.
//!
//! What's NOT in this template by default:
//! * a concrete file/SQLite/HTTP `TaskStore` implementation
//! * `cancel` / `list` / `download`
//! * task creation flows
//!
//! Those are domain-specific and easy to write once you know the trait.
//! See `templates/RECIPES.md` in the parent agent-cli-builder skill for
//! worked examples (file-backed store, cancel + list, download).

use std::time::Duration;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::errors::CliError;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TaskState {
    Pending,
    Running,
    Succeeded,
    Failed,
    Cancelled,
}

impl TaskState {
    #[must_use]
    pub const fn is_terminal(self) -> bool {
        matches!(self, Self::Succeeded | Self::Failed | Self::Cancelled)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Task {
    pub task_id: String,
    pub state: TaskState,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub kind: String,
    pub payload: serde_json::Value,
    pub error: Option<String>,
}

/// The task storage contract. Implement this against your real backend.
/// `mycli task get` and `mycli task wait` only need `get`. Add `cancel`
/// and `list` when you need them; see the recipes file in the parent skill.
#[allow(async_fn_in_trait)]
pub trait TaskStore {
    async fn get(&self, id: &str) -> Result<Task, CliError>;
}

/// Block until the task reaches a terminal state or `timeout` elapses.
/// Returns the final task on success; on timeout, exit code 5 (TIMEOUT)
/// so agents can branch on `$?` without parsing strings.
pub async fn wait_for_terminal<S: TaskStore>(
    store: &S,
    id: &str,
    timeout: Duration,
    poll_interval: Duration,
) -> Result<Task, CliError> {
    let start = std::time::Instant::now();
    loop {
        let task = store.get(id).await?;
        if task.state.is_terminal() {
            return Ok(task);
        }
        if start.elapsed() >= timeout {
            return Err(CliError::timeout(format!(
                "task {id} still {:?} after {}s; call `mycli task get {id}` to re-check",
                task.state,
                timeout.as_secs()
            )));
        }
        tokio::time::sleep(poll_interval).await;
    }
}
